### DISCUSSION TITLE
`pie-icons-webc` approach

### DISCUSSION THREAD

**Discussion by ashleynolan** [justeattakeaway/pie#531] [Ideas]
Just wanted to spin up a discussion thread ahead of the planned work to create an icon package which will support Web Components. Hopefully, this should cover off some of the reasoning and open up discussion around the potential approach.

## Why do we need this?

Currently, we have `pie-icons-react` and `pie-icons-vue`. These packages compile the icons in `pie-icons` out into React of Vue components, while also exposing any props to help with the use of these icons – the main one to provide a way of being able to specify the size of the icon against PIE standards (so for example, sizing like `s`/`m`/`l` converting to width/height on the SVG automatically).

To provide the same sizing helpers for web components, we'll need to similarly compile out our SVGs to Lit to provide these helper properties. It makes sense to do this with Lit, as this is the library we are using for our other web components.

## Proposed approach

Currently, our assumption is that we would just copy the React/Vue packages in terms of their interface, so providing an `iconSize` property that would set the width/height of the icon automatically on the SVG. The `iconSize` prop would work in the same way as it does for the react and vue icon packages currently.

Consumers would need to specify between using different Icon variants (for example, most icons have a regular and large variant). We _could_ provide a nicer way of consuming icons with variations, but I suspect that this may mean that components increase in size, so although using the icon from a dev point of view would be nicer, it may have performance implications if consuming apps aren't able to tree-shake that code away.

## Considerations

- SSR – Icons rendered as a Web Component don't automatically get rendered server side. Platforms would need to ensure they use the appropriate SSR plugins for their platform so that icons are visible server side. This is likely not a huge issue for most icons where text fallbacks are available, but for any icon that is the sole icon being used (for example `pie-icon-button`) this needs to be considered.
- Is this the best approach? (Any considerations here welcome!)

**Comment by thejfreitas** [comment]
Hello.

The variants might be a little challenging considering different scenarios. Look at the images below:
 
<img width="1547" alt="Screenshot 2023-06-12 at 5 32 34 PM" src="https://github.com/justeattakeaway/pie/assets/7768202/86671923-f804-471b-90d2-56ae787c0e38">
<img width="1165" alt="Screenshot 2023-06-12 at 5 31 53 PM" src="https://github.com/justeattakeaway/pie/assets/7768202/81b8d6d7-a786-4b1a-a3d5-8c4a77f95e1a">

Some components looks pretty much the same but their borders are thicker  and some has differences. This way there are different SVGs for each variant. In this case we may create a component like mentioned above `<pie-icon-button variant="large">` and the user can fetch the component on demand but each component will carry two different SVGs.

Perhaps another option (_Probably not the best_) is to create a single component that renders that icon according to the property declared like: `<pie-icon icon="ArrowLongLeft">` or `<pie-icon icon="ArrowLongLeftLarge">` however I think this way will be too expensive to all SVGs in only one component.

**Reply by jamieomaguire** [reply]
@thejfreitas I agree that the variants present a challenge, that's a really good point. 

I really like the idea of `<pie-icon icon="ArrowLongLeft">` however as you mention I am concerned about the performance impact this could have on consuming applications. I don't know if there is any kind of dynamic fetching that could be done to mitigate this. Perhaps this is something we could look into?

The current pattern of having 1 component per icon variant seems to have been accepted across the org already. So sticking with that approach, for now, could be the simplest approach.

One minor nitpick: could we use `size` instead of `iconSize`? It's more readable and I think the word 'icon' may be a little redundant given the context.

**Reply by ashleynolan** [reply]
I've no problem with us moving to just use `size` instead of `iconSize`.

I agree on the variants as well – I think it could cause more issues for users than it solves. Some icons have large variants, some don't. Some have filled variants, some don't.

I think because there isn't a consistent set of variants it may be better to just let users of the icon package directly specify the icon they want to use. It's slightly more specific, but at least it avoids the potential confusion of 'Why doesn't this icon have a large/filled variant?'

**Reply by thejfreitas** [reply]
> @thejfreitas I agree that the variants present a challenge, that's a really good point.
> 
> I really like the idea of `<pie-icon icon="ArrowLongLeft">` however as you mention I am concerned about the performance impact this could have on consuming applications. I don't know if there is any kind of dynamic fetching that could be done to mitigate this. Perhaps this is something we could look into?
> 
> The current pattern of having 1 component per icon variant seems to have been accepted across the org already. So sticking with that approach, for now, could be the simplest approach.
> 
> One minor nitpick: could we use `size` instead of `iconSize`? It's more readable and I think the word 'icon' may be a little redundant given the context.

I like the idea to use `size` instead of `iconSize`

**Reply by xander-marjoram** [reply]
I thought I would just add that the vast majority of the icons do have regular and large variants. It's only the flags and payment icons that have a single size.

**Comment by davidpn11** [comment]
I think the problem with using a component in the `<pie-icon icon="ArrowLongLeft">` format is that it will be hard to code split the icons, as we only have one component that can load all the icons. 

The way we currently use the icons is fully based on the `<svg/>` tag API, so use `height` and `width` for sizes and `fill` for color. I do like that approach as it is the most common, but I do understand simplifying by extending it to a `size` property. Would we set the `size` prop with strings or numbers? 

```
size="10px"
//or
size={10}
```

Also, since we are discussing extending the API to support the `size`, should we also do it for "color", maybe something attached to our PIE tokens and themes? 
Not sure this is the way to go as it can add more complexity to the components, but would be nice to include it in the discussion

**Reply by jamieomaguire** [reply]
I'm interested to know how granular we need to allow `size` to be - can we predefine some sizes and label them as `s`, `m`, `l`, for example? I.e. `size="m"`.

I'm sure there are some situations where we might need to override with a particular pixel size but I haven't come across many cases for this yet. This could make bigger branding changes easier in the future too as we only need to change the mapping between the size label and pixel value.

**Reply by ashleynolan** [reply]
So there are currently two sizes defined by design for icon usage. `regular` icons (icons not suffixed with `Large` can be:

```js
{
    xs: 16,
    s: 20,
    m: 24,
    l: 28,
    xl: 32,
    xxl: 40
}
```

Large icons must be a minimum of 32px width/height and can be any size bigger than that as long as it's divisible by 8 (so 40px, 48px, 80px, 96px are all valid sizes for large icons).

In terms of colour – I don't think we'll need to have a prop for it. We've recently stripped out fill colours in favour of using `fill: currentColor` which is what most icon libraries use. That way, the fill can be set by the consuming application easily by just targeting the component (users can either set `fill` or `color` on the element and the icon will take that as the colour). We don't have any strict colours for icons, only guidance.

Here's the [link to the PIE Icons docs in Figma](https://www.figma.com/file/k7gPJ4MZRUj4nlZK2hL0Op/%5BCore%5D-Icons-%5BPIE-2.0%5D?type=design&node-id=1-52&t=IYhmz6GUEchiUp0Z-0) which has the info on the icon size and Colours

**Comment by ashleynolan** [comment]
I'm going to mark this discussion as closed, as I think it captures all the requirements we needed covering for the new package. Thanks for the comments and feedback – when we have something to PR, would be good to get reviews of that!
